dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/datafile/FileIntegrityChecker.java

```diff
@@ -0,0 +1,131 @@
+            DataAccess dataAccess) {
+        this.dataFileService = dataFileServiceBean;
+        this.authSvc = authSvc;
+        this.mailService = mailService;
+        this.dataAccess = dataAccess;
+    }
+    
+    
+    // -------------------- LOGIC --------------------
+
+    public FilesIntegrityReport checkFilesIntegrity() {
+        List<DataFile> dataFiles = dataFileService.findAll();
+
+        FilesIntegrityReport report = new FilesIntegrityReport();
+        report.setCheckedCount(dataFiles.size());
+
+        for (DataFile dataFile:dataFiles) {
+            FileIntegrityCheckResult checkResult = checkFileIntegrity(dataFile);
+
+            if (!checkResult.isOK()) {
+                report.addSuspicious(dataFile, checkResult);
+            } else if (checkResult == FileIntegrityCheckResult.OK_SKIPPED_CHECKSUM_VERIFICATION) {
+                report.incrementSkippedChecksumVerification();
+            }
+        }
+
+        EmailContent reportEmailContent = buildReportEmailContent(report);
+        authSvc.findSuperUsers()
+            .stream()
+            .forEach(user -> mailService.sendMailAsync(user.getEmail(), reportEmailContent));
+
+
+        return report;
+
+    }
+
+    // -------------------- PRIVATE --------------------
+
+    private FileIntegrityCheckResult checkFileIntegrity(DataFile dataFile) {
+        try {
+            StorageIO<DataFile> storageIO = dataAccess.getStorageIO(dataFile);
+            
+            if (!storageIO.exists() || storageIO.getSize() == 0) {
+                return FileIntegrityCheckResult.NOT_EXIST;
+            }
+            if (storageIO.getSize() != dataFile.getFilesize()) {
+                return FileIntegrityCheckResult.DIFFERENT_SIZE;
+            }
+            if (storageIO.isMD5CheckSupported() && dataFile.getChecksumType() == ChecksumType.MD5) {
+                if (!StringUtils.equals(storageIO.getMD5(), dataFile.getChecksumValue())) {
+                    return FileIntegrityCheckResult.DIFFERENT_CHECKSUM;
+                }
+            } else {
+                return FileIntegrityCheckResult.OK_SKIPPED_CHECKSUM_VERIFICATION;
+            }
+            return FileIntegrityCheckResult.OK;
+            
+        } catch (IOException e) {
+            logger.info(e.getMessage());
+            return FileIntegrityCheckResult.STORAGE_ERROR;
+        }
+    }
+
+    private EmailContent buildReportEmailContent(FilesIntegrityReport report) {
+        StringBuilder messageBodyBuilder = new StringBuilder();
+        
+        messageBodyBuilder.append("Datafiles integrity check summary: \n");
+        messageBodyBuilder.append("Files checked: " + report.getCheckedCount() + "\n");
+        messageBodyBuilder.append("Skipped checksum verification: " + report.getSkippedChecksumVerification() + "\n");
+        messageBodyBuilder.append("Number of files with failures: " + report.getSuspicious().size() + "\n\n");
+        messageBodyBuilder.append("List of files with failures:\n");
```

Comment:
Maybe we can use `append` instead of `+`, as the latter creates another `StringBuilder` in the background. Also we can chain all `append` calls.

Reply:
I don't think it makes much difference - but ok.